[QENG-3919] Make vmpooler checkouts be all or nothing

rick opened the pull request:
![](https://upload.wikimedia.org/wikipedia/en/d/de/Jay_Sean_-_All_or_Nothing.jpg)

Prior to this, vmpooler could partially fill requests for multiple VMs. Here we move to an algorithm that returns the entire set of requested VMs or no VMs.

/cc @puppetlabs/cinext

shermdog commented:
Do we consider this an API v2 type thing?

waynr commented:
@shermdog my understanding is that this is addressing a bug in the v1 API so I don't think so.

briancain commented:
Does this deprecate the use of fetching vms through query params like what was mentioned in the ticket, or no? I don't see any spec tests around that, just fetching vms through a request body.

rick commented:
Good eye. I think we should probably bolster tests around that query params
use case.

On Mon, May 23, 2016 at 4:56 PM Brian Cain <notifications@github.com> wrote:

> Does this deprecate the use of fetching vms through query params like what
> was mentioned in the ticket, or no? I don't see any spec tests around that,
> just fetching vms through a request body.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/vmpooler/pull/153#issuecomment-221108090>
>

rick commented:
Note to self: also, double-check the docs to make sure they're clear about the intended behavior.

rick commented:
Would actually be good to get another look at this, as I've added tests for and fixed the second code path, refactored some things and tweaked the docs.

waynr commented on lib/vmpooler/api/v1.rb:
```diff
@@ -45,38 +45,88 @@ def alias_deref(hash)
+        account_for_starting_vm(template, vm)
+        update_result_hosts(result, template, vm)
+      else
+        status 503
```
In the specs I see a comment that says "which HTTP status code?" on tests that seem (by the name) to be expecting failure and I also see `http = 200` in the `expect_json` call. Should that be `http = 503`?

rick replied:
I added those comments and will clean them up (thanks for the reminder) ... It looks like vmpooler has always returned a 200 status, even when it cannot allocate vm's, using `return['ok']` as the indicator of whether the allocation was ok.  The 503 status is for cases where no templates were specified.

I don't 100% agree with the way it's using status + `result['ok']`, but I'm going to leave it alone, otherwise this would be a breaking change for folks.

rick replied:
Dropped the comments in question after another round of verifying behavior.

waynr commented on lib/vmpooler/api/v1.rb:
```diff
@@ -45,38 +45,88 @@ def alias_deref(hash)
-          backend.hset('vmpooler__vm__' + vm, 'token:user',
-            backend.hget('vmpooler__token__' + request.env['HTTP_X_AUTH_TOKEN'], 'user')
-          )
+    def atomically_allocate_vms(payload)
```
In the case of single-vm requests we "checkout vm" but in the case of multiple-vm requests we "atomically allocate vms"...should we try to align the semantics between these two methods to make it clear that they serve similar purposes? (the former being an operation that returns a scalar whereas the latter returns a collection) At the very least it seems like we should be able to say that the scalar operation is trivially atomic.

rick replied:
Turns out, `checkout_vm` is no longer even used any more. :fire: :skull:

waynr commented on spec/vmpooler/api/v1_spec.rb:
```diff
@@ -417,6 +531,120 @@ def app()
+      end
+
+      it 'returns multiple VMs even when multiple instances from multiple pools are requested' do
+        post "#{prefix}/vm/pool1+pool1+pool2+pool2+pool2", ''
```
I could be missing something, but it seems like the only difference between this new block of test and the previous new block of tests is the URL. Does RSpec allow for a test like this to be parameterized so you can tell the test case to run with each valid URL that is expected to produce the same results?

rick replied:
It's possible to do something along those lines, but it would basically be parametrizing pairs of tests across a 6 test set, which tends to look more confusing than just having the duplication in the tests.

rick commented:
Going to go over the return status stuff one more time. I think I dropped the '503' on the floor in this change and want to make sure I get these right.

rick commented:
Trivia fun facts:

 - To test this stuff locally, you probably want to grab a database dump from like vmpooler-dev and install it in your local redis
 - You're going to get `{ "ok" => "false" }` results on everything unless you can auth. That means making a valid token on any test instances.

Testing results on vmpooler-dev:

```
$ curl -X POST -u rick.bradley --url vmpooler-dev.delivery.puppetlabs.net/api/v1/token -d ''
Thu May 26 15:53:45 chenrezig ~/pl/vmpooler
 (qeng-3919-all-or-nothing-checkout *%) % ▸ curl --header "X-AUTH-TOKEN: xxxx" -X POST -d '' --url vmpooler-dev.delivery.puppetlabs.net/api/v1/vm/centos-6-x86_64+centos-6-x86_64+centos-6-x86_64+centos-6-x86_64+centos-6-x86_64+centos-6-x86_64
{
  "ok": false
}Thu May 26 15:53:56 chenrezig ~/pl/vmpooler
 (qeng-3919-all-or-nothing-checkout *%) % ▸ curl --header "X-AUTH-TOKEN: xxxx" -X POST -d '' --url vmpooler-dev.delivery.puppetlabs.net/api/v1/vm/centos-6-x86_64+centos-6-x86_64+centos-6-x86_64+centos-6-x86_64
{
  "ok": true,
  "centos-6-x86_64": {
    "hostname": [
      "n21nif8dtxkbt8o",
      "b0v7kcczu4iuxwd",
      "waobk9n3d5j5ljr",
      "pcxn7yx8i6awz86"
    ]
  },
  "domain": "delivery.puppetlabs.net"
}Thu May 26 15:54:45 chenrezig ~/pl/vmpooler
 (qeng-3919-all-or-nothing-checkout *%) % ▸ curl --header "X-AUTH-TOKEN: xxxx" --url vmpooler-dev.delivery.puppetlabs.net/api/v1/vm/
[
  "centos-6-x86_64",
  "debian-7-i386",
  "debian-7-x86_64",
  "osx-1010-x86_64",
  "sles-10-x86_64",
  "solaris-10-x86_64",
  "win-2008r2-x86_64"
]Thu May 26 15:55:09 chenrezig ~/pl/vmpooler
 (qeng-3919-all-or-nothing-checkout *%) % ▸ curl --header "X-AUTH-TOKEN: xxxx" -X POST -d '' --url vmpooler-dev.delivery.puppetlabs.net/api/v1/vm/centos-6-x86_64+centos-6-x86_64+centos-6-x86_64+centos-6-x86_64+osx-1010-x86_64+osx-1010-x86_64
{
  "ok": false
}Thu May 26 15:55:40 chenrezig ~/pl/vmpooler
 (qeng-3919-all-or-nothing-checkout *%) % ▸ curl --header "X-AUTH-TOKEN: xxxx" -X POST -d '' --url vmpooler-dev.delivery.puppetlabs.net/api/v1/vm/debian-7-i386+debian-7-i386+debian-7-i386+osx-1010-x86_64+osx-1010-x86_64
{
  "ok": true,
  "debian-7-i386": {
    "hostname": [
      "spa053w9pu64jx8",
      "wpnlypog5wkq9gg",
      "fayw2046exh8icn"
    ]
  },
  "osx-1010-x86_64": {
    "hostname": [
      "uus0eaqy8we41qr",
      "pbjbfqawk0vktko"
    ]
  },
  "domain": "delivery.puppetlabs.net"
}Thu May 26 15:56:21 chenrezig ~/pl/vmpooler
 (qeng-3919-all-or-nothing-checkout *%) % ▸ curl --header "X-AUTH-TOKEN: xxxx" -X POST -d '' --url vmpooler-dev.delivery.puppetlabs.net/api/v1/vm/debian-7-i386+debian-7-i386+debian-7-i386+osx-1010-x86_64+osx-1010-x86_64
{
  "ok": false
}
Thu May 26 15:56:24 chenrezig ~/pl/vmpooler

```

waynr commented:
@rick it seems like those development setup gotchas should be documented somewhere, should we include that in this PR or file a ticket to write up a more thorough guide to developing on vmpooler?

rick commented:
@waynr yeah, no doubt.  There's a part of it (maybe specific examples) that are pretty specific to our internal network ("here, do this to grab a db dump and throw it in your local db, and then make a token and do this"), so perhaps for some of our internal wiki docs we get really specific, and for vmpooler in general we document the more general process.

rick commented:
As an aside, after testing this in dev I think this is good to :ship:, but one last :eyes: wouldn't hurt.

rick commented:
(this is also a great candidate for github's new squash merge) :tada:

rick commented on Gemfile:
```diff
@@ -6,6 +6,7 @@ gem 'rake', '>= 10.4'
 gem 'rbvmomi', '>= 1.8'
 gem 'redis', '>= 3.2'
 gem 'sinatra', '>= 1.4'
+gem 'net-ldap', '= 0.11'
```
cc @shermdog (hard-wired to make travis happy on jruby, etc.)

shermdog replied:
Let's not hardwire as we do have some customers using this tool.

waynr commented:
Looks good to me! :+1:

shermdog commented on lib/vmpooler/api/v1.rb:
```diff
@@ -45,38 +45,75 @@ def alias_deref(hash)
+      backend.spop('vmpooler__ready__' + template)
+    end
+
+    def return_single_vm(template, vm)
```
This confused me at first - could we rename it to something like re_ready_single_vm or something along those lines?

rick replied:
Sure thing!
